src/civis/utils/_jobs.py

```
@@ -96,3 +104,143 @@ def run_template(id, arguments, JSONValue=False, client=None):
+        return sorted_logs[-_LOG_REFETCH_COUNT]["id"]
+
+    return 0
+
+
+def _job_finished_past_timeout(job_id, run_id, finished_timeout, raw_client):
+    """Return true if the run finished more than so many seconds ago."""
+    if finished_timeout is None:
+        return False
+
+    run = raw_client.jobs.get_runs(job_id, run_id)
+    finished_at = run.json()["finishedAt"]
+    if finished_at is None:
+        return False
+    finished_at_ts = _timestamp_from_iso_str(finished_at)
+    result = finished_at_ts < time.time() - finished_timeout
+    return result
+
+
+def job_logs(job_id, run_id=None, raw_client=None, finished_timeout=None):
+    """Return a generator of log message dictionaries for a given run.
+
+    Parameters
+    ----------
+    job_id : int
+        The ID of the job to retrieve log message for.
+    run_id : int or None
+        The ID of the run to retrieve log messages for.
+        If None, the ID for the most recent run will be used.
+    raw_client: :class:`civis.APIClient`, optional
+        If not provided, an :class:`civis.APIClient` object will be
+        created from the :envvar:`CIVIS_API_KEY`.
+        The return_type should be set to "raw".
+    finished_timeout: int or None
+        If not None, then this function will return once the run has
+        been finished for the specified number of seconds.
+        If None, then this function will wait until the API says there
+        will be no more new log messages, which may take a few minutes.
+        A timeout of 30-60 seconds is usually enough to retrieve all
+        log messages.
+
+    Yields
+    ------
+    dict
+        A log message dictionary with "message", "createdAt" and other attributes
+        provided by the job logs endpoint. Note that this will block execution
+        until the job has stopped and all log messages are retrieved.
+    """
+    if raw_client is None:
+        raw_client = APIClient(return_type="raw")
```

This requirement of a "raw" client seems odd. Is it possible to update this function to use the regular default / "snake" client instead? I know this "raw" client comes from the previously internal-only code , but now that we're exposing this argument in a public function like this, it would seem odd for this and only this place to require a "raw" client, while everywhere else in civis-python the regular "snake" client is required. To use this function as-is while needing to explicitly pass in an API client object (e.g., when handling both robot and non-robot API keys in the same execution environment -- not too uncommon these days in internal work), one would have to instantiate two client objects for a given API key, the "raw" one just for this function and the regular one for everything else, which would be strange.

The default "snake" `client` would get us a list of `civis.Response` objects returned by `client.jobs.list_runs_logs(...)`. The `headers` dict is [available at an attribute by the same name](https://github.com/civisanalytics/civis-python/blob/d5153897914ed799164637672bd18a2492a023f1/src/civis/response.py#L132) at the first `civis.Response` object in this list (all objects in this list [has the same `headers` content](https://github.com/civisanalytics/civis-python/blob/d5153897914ed799164637672bd18a2492a023f1/src/civis/response.py#L92-L96)). When looping through each of these response objects, calling `.json()` on this object would return [a dict with "snake-case" keys by default](https://github.com/civisanalytics/civis-python/blob/d5153897914ed799164637672bd18a2492a023f1/src/civis/response.py#L169), which has the added benefit of using/expecting "created_at" not "createdAt" etc. to conform to Python's preference for snake-case styling.

---

Oh, I didn't realize `headers` was part of the `Response` object. Perhaps it wasn't when this code was originally written.  I'll give that a try.

---

Hmm, it looks like the normal API client will return an empty list without the headers if there are no new logs, so the raw client is needed to check `civis-cache-control` when that happens. I don't immediately see a way around this. Thoughts?

---

I see! How about if the user provides a regular "snake" client, the function internally creates a "raw" one based on the user-provided one? I was only 50-50 for whether this approach might work, so I wanted to be more certain -- [this](https://github.com/civisanalytics/civis-python/compare/add-log-generation-function...add-log-generation-function-use-regular-client) seems to work?

If you think this direction is a no-go, i.e., if a user-provided client must be a "raw" one, I'd recommend the following:
* Name the argument `client` instead of `raw_client`, to anticipate that we may one day be able to allow a user-provided regular "snake" client in a future release, without having to deal with a breaking change from `raw_client` to `client`.
* If a "raw" client is required, it should be formally checked (in addition to documentation) and we should raise an exception if this requirement is not met (in which case we likely still would need to keep a `self._return_type` directly at a client object for this purpose).

---

Hmm... I'm concerned there might be issues if someone used multithreading with this approach where `_return_type` is changed and then changed back. We could add a lock of some kind, but that seems overly complex.

The main reason I was passing in `raw_client` was to avoid an extra API call to `/endpoints`, but `generate_classes_maybe_cached` should help avoid that. How about I just not have a `client` or `raw_client` argument, which is how the CLI command was before this?

---

OK. I removed the arg in [41c381a](https://github.com/civisanalytics/civis-python/pull/509/commits/41c381a0a9b808c1e6709783ac8c2e8e8a680027)